### DISCUSSION TITLE
React 18 & React-Testing-Library performance vs Jest

### DISCUSSION THREAD

**Discussion by is-jonreeves** [vitest-dev/vitest#6098] [General]
I've been following `vitest` for over a year and up till recently I had only used it in a few node projects. I have been really impressed with the performance so decided to adopt it for a bigger React 18 SPA I work on. I was surprised to see the performance vs `jest` take quite a hit.

We have ~1100 tests and locally with `jest` (CRA + Craco + Babel) they took about 90s on average on my laptop, 120s in CI. After moving to `vite`, and then adopting `vitest` I now get 200s locally and 240s in CI.

I know that these topics are less than helpful without examples, but I'm unfortunately unable to share this code base. I was really just raising this topic to see if what I'm seeing is out of the norm (and most people are getting the same or better timings vs jest), or if vitest is a little slower in this context right now?

I have tried both `happy-dom` and `jsdom`, as well as `forks` and `threads`, and get similar results. I haven't spent a great deal of time debugging the tests yet to see if something in particular is taking longer than I would expect. But skimming over their individual timings, they don't seem unusual at first glance, and from what I can see my CPU is pegged at 100% throughout the run.

During the migration I did run into two issues that might be impacting things (or indicators of something wrong in some tests)...

the first was `v8` coverage not working for me due to hanging processes, when trying to debug getting would just get empty stack traces. Usually it would surface the following error, so I reverted back to `istanbul` for now. Either way the timings above are with coverage disabled and the only reporter set to `verbose` (so is unlikely impacting times).
```
TypeError: undefined is not iterable (cannot read property Symbol(Symbol.iterator))
 ❯ RangeTree.fromSortedRanges node_modules/@bcoe/v8-coverage/dist/lib/range-tree.js:27:39
 ❯ normalizeFunctionCov node_modules/@bcoe/v8-coverage/dist/lib/normalize.js:74:41
 ❯ Object.deepNormalizeScriptCov node_modules/@bcoe/v8-coverage/dist/lib/normalize.js:59:9
 ❯ mergeScriptCovs node_modules/@bcoe/v8-coverage/dist/lib/merge.js:59:21
 ❯ mergeProcessCovs node_modules/@bcoe/v8-coverage/dist/lib/merge.js:34:21
 ❯ node_modules/@vitest/coverage-v8/dist/provider.js:2375:24
 ❯ V8CoverageProvider.generateCoverage node_modules/@vitest/coverage-v8/dist/provider.js:2371:11
 ❯ node_modules/vitest/dist/vendor/cli-api.BGRtK_Kj.js:14768:24
 ❯ Vitest.runFiles node_modules/vitest/dist/vendor/cli-api.BGRtK_Kj.js:14776:12
 ❯ Vitest.start node_modules/vitest/dist/vendor/cli-api.BGRtK_Kj.js:14622:7
 ❯ startVitest node_modules/vitest/dist/vendor/cli-api.BGRtK_Kj.js:15729:7
 ❯ start node_modules/vitest/dist/vendor/cac.BQc6bsef.js:1502:17
 ❯ CAC.run node_modules/vitest/dist/vendor/cac.BQc6bsef.js:1482:3
```

the second issue was that `vi.useFakeTimers()` didn't seem to play well with `@testing-library/react`. To get `vi.advanceTimersByTime()` to work, I had to use `vi.useFakeTimers({ shouldAdvanceTime: true })` which didn't feel right, so am wondering if its synthetically delaying stuff. There are some issues [here](https://github.com/testing-library/react-testing-library/issues/1198), [here](https://github.com/testing-library/react-testing-library/issues/1197) and a [workaround suggestion here](https://github.com/testing-library/user-event/issues/1115#issuecomment-1565730917).

I plan to spend some time next week debugging further, but was hoping to get an idea from others about what performance differences they are seeing between jest and vitest in their projects, so I can set my expectations.

**Comment by sheremet-va** [comment]
Don't have much time to answer everything here, but for performance issues you can try using `vmForks` which should be close in performance to jest (that's its only purpose)

**Reply by is-jonreeves** [reply]
thanks for the suggestion, I gave it a go but the numbers are pretty similar in my case:
```
# threads
218.03s (transform 71.04s, setup 216.15s, collect 1556.05s, tests 150.26s, environment 200.78s, prepare 86.90s)
215.78s (transform 70.73s, setup 209.97s, collect 1548.45s, tests 150.26s, environment 198.17s, prepare 86.13s)
214.54s (transform 69.17s, setup 208.83s, collect 1527.84s, tests 147.57s, environment 197.62s, prepare 82.17s)

# vmForks
234.70s (transform 60.81s, setup 89.22s, collect 2136.41s, tests 126.42s, environment 16.99s, prepare 110.27s)
225.84s (transform 56.16s, setup 87.41s, collect 2053.75s, tests 122.06s, environment 15.61s, prepare 104.46s)
230.60s (transform 56.54s, setup 86.80s, collect 2099.12s, tests 127.19s, environment 15.32s, prepare 106.07s)
```

When using `vmForks` I had 4 tests fail (which I'll need to investigate) which may account for the additional ~15-20s (due to timeouts). I did also notice an occasional flash of this message:
```
MaxListenersExceededWarning: Possible EventEmmitter memory leak detected. 11 message listeners added to [process]. Use emitter.setMaxListeners() to increase limit
```

Will make a note to explore further with `--trace-warnings` and see if there are some problematic tests.

**Reply by iwan-uschka** [reply]
I just ran into performance issues regarding memory consumption using `vmForks`. Jobs in CI environment were failing due to lack of memory (vitest needed +6GB). The thing about `vmForks` seems to be that memory usage increases during runtime and you can't do nothing about it. To be sure about that i just created a suite of 1.044.000 (equal) tests testing the simple function

```TypeScript
const functionToTest = (value = false) => {
  return value;
};
```

Using `vmForks` and `isolate: false` led to an increasing memory consumption up to 700 Mbyte for each process, while running 14 processes in parallel.

Using `forks` and `isolate: false` led to a memory consumption increasing up to 180 MByte and falling back to 40 MByte per process, while running 8 processes in parallel.

Using `forks` and `isolate: true` (defaults of vitest) led to a constant memory consumption of 45 MByte per process (one process per test i guess). But: the runtime went up from 30 seconds to 105 seconds.

**Reply by maxgruebneraeroqual** [reply]
Thought I'd add a comment for anyone else stumbling across this thread.
We swapped from jest to vitest for a suite of a few hundred react tests in a reasonably sized modernish React application. I've found it difficult to find examples of people using this with large codebases.
The default test run for vitest took about 10 minutes. 
Disabling isolation was faster, but breaks heaps of tests because of how the react tests share resources in jsdom (I guess?)
Switching to vmforks, which isn't particularly well documented in the 'improving-performance' docs, dropped that to about 2 minutes, but also consumed 16gb RAM, which made it unsuitable for our CI system.

**Comment by is-jonreeves** [comment]
I moved away from using `vi.useFakeTimers({ shouldAdvanceTime: true })`, and instead used the aforementioned workaround in my setup file:
```ts
beforeEach(() => {
  // see https://github.com/testing-library/user-event/issues/1115#issuecomment-1565730917
  if ('vi' in global) vi.stubGlobal('jest', { advanceTimersByTime: vi.advanceTimersByTime.bind(vi) });
});
```

This change did not affect my timings, so `shouldAdvanceTime` likely was not impacting things.

I also noted a few tests where `happy-dom` was attempting a network request, so thought that might be affecting things. I resolved that within the specific test suites with:
```ts
beforeAll(() => {
  // see https://github.com/capricorn86/happy-dom/wiki/IOptionalBrowserSettings
  vi.stubGlobal('happyDOM', merge(('happyDOM' in global) ? global.happyDOM : {}, { settings: { disableJavaScriptFileLoading: true, handleDisabledFileLoadingAsSuccess: true } }));
});

afterAll(() => {
  vi.unstubAllGlobals();
});
```

This change also did not affect my timings.

**Comment by is-jonreeves** [comment]
After reading through a few of the older issues, including #579 I think I have a pretty good idea about what to expect and what tuning is available to me. This article was linked in that issue and has helped: https://dev.to/thejaredwilcurt/improving-vitest-performance-42c6

The following actions had the greatest impact in my case. The tests are now running in 90-100s which is close enough to what we had in `jest`:
- using `vmForks`
- using `deps.optimizer.web.include` (icons, sdks, component-libs)
- using a [plugin](https://github.com/vitest-dev/vitest/issues/579#issuecomment-1197227773) to for `.module.scss` files
